### ISSUE TITLE
a tool for asynchronous processing of proof files

### ISSUE THREAD

**Issue by DanGrayson** [UniMath/UniMath#780]
Yesterday I spoke to Karl Palmskog <palmskog@illinois.edu> about the paper

https://www.cs.utexas.edu/~celik/papers/CelikETAL17iCoq.pdf

and how it might help speed up processing of proof files by `coqc` by taking full
advantage of a recent feature of Coq for checking proofs asynchronously.  Their code
is at

https://github.com/ProofEngineering

, and there you can find one of the key plug-ins:

https://github.com/proofengineering/coq-depends

Take a look at

https://coq.inria.fr/refman/async-proofs.html

and you see that opaque proofs can be checked independently, which one might
expect, because the proof is not needed for later processing.

The main idea of their tools is to compute a fine-grained 
dependency graph whose nodes are individual
proofs -- currently we build *.vo files by letting `make` produce a dependency graph
whose nodes are the *.v files.  This allows them to produce *.vio files using the `-quick` option
of `coqc` -- those contain the types of the definitions but not the terms, since no proof checking is done.  Then their tool figures out which proofs need to be rechecked, and when.  Those steps
be run in parallel on several cores.

I'm not sure how that all works when running Coq interactively.

There might be a problem with polymorphic universe parameters, with not all of the 
universe constraints being properly detected.  I assume that's because the new 
constraints imposed on the universe level parameters associated with a definition 
depend on the proof term.  If so, we wouldn't want to use it for Travis.  (We will start
using polymorphic universe parameters once we are using type-in-type on just the
resizing axioms.)

They did the following test.  They ran through a few dozen of our pull requests sequentially,
caching the *.vio files from each one for use in the next (we can do the same in travis), and
got a speedup factor of 3.5.  Making many more definitions opaque might improve that to
a factor of 10.

One of the things needed to make things go optimally is the addition of annotations to each
proof in a section that specifies which of the section-wide assumptions will be used.  That
allows the "real" type of the resulting definition to be determined, as then those assumptions
will be the added parameters.  They also have a tool for figuring out what text to add to
each `Proof.` line; it involves the `using` parameter: see 7.1.5 in the Coq manual.  He
is willing to make a pull request for us that adds the right thing to each proof in a section.

We should learn enough about this to determine whether using their tools can speed up
our checking time in Travis, and whether it can speed up our checking time at home, without
losing certainty about universe consistency.

**Comment by arnoudvanderleer** [comment]
@DanGrayson Is this still a thing that we would want? Or is this now already covered by faster computers / dune compilation / new features in coq itself?

**Comment by benediktahrens** [comment]
This is more a report on what is (was) happening in the Coq eco-system, not a real issue for UniMath. A better place for discussions like this would be the UniMath Zulip, which we did not have in 2017. I will close this issue.

Btw, I believe such processing is now done in VSCoq 2?

**Comment by nmvdw** [comment]
> Btw, I believe such processing is now done in VSCoq 2?

Yes. In the newest version of VSCoq, one can choose to use that.
